File: sln/test/NSpecSpecs/describe_RunningSpecs/describe_example.cs
using System;
using NSpec;
using NUnit.Framework;
using NSpecSpecs.describe_RunningSpecs.Exceptions;
using FluentAssertions;

namespace NSpecSpecs.WhenRunningSpecs
{
    [TestFixture]
    [Category("RunningSpecs")]
    public class describe_example : when_running_specs
    {
        class SpecClass : nspec
        {
            void it_changes_status_after_run()
            {
            }

            void it_passes()
            {
            }

            void it_fails()
            {
                throw new KnownException();
            }
        }

        [Test]
        public void execution_status_changes_after_run()
        {
            Run(typeof(SpecClass));

            var ex = TheExample("it changes status after run");

            //ex.HasRun.Should().BeFalse(); //broken after making init and run happen all at once

            ex.HasRun.Should().BeTrue();
        }

        [Test]
        public void duration_is_set_after_run()
        {
            Run(typeof(SpecClass));

            var ex = TheExample("it changes status after run");

            ex.Duration.Should().BeGreaterThan(TimeSpan.Zero);
        }

        [Test]
        public void passing_status_is_passed_when_it_succeeds()
        {
            Run(typeof(SpecClass));

            TheExample("it passes").ShouldHavePassed();
        }

        [Test]
        public void passing_status_is_not_passed_when_it_fails()
        {
            Run(typeof(SpecClass));

            TheExample("it fails").ShouldHaveFailed();
        }

        class SpecClassWithAnonymousLambdas : nspec
        {
            void describe_specs_with_anonymous_lambdas()
            {
                context["Some context with anonymous lambdas"] = () =>
                {
                    it["has an anonymous lambda"] = () =>
                    {
                    };
                };
            }
        }

        [Test]
        public void finds_and_runs_three_class_level_examples()
        {
            Run(typeof(SpecClass));

            TheExampleCount().Should().Be(3);
        }

        [Test]
        public void finds_and_runs_only_one_example_ignoring_anonymous_lambdas()
        {
            Run(typeof(SpecClassWithAnonymousLambdas));

            TheExampleCount().Should().Be(1);
        }
    }
}

File: sln/test/NSpec.Tests.VB/Describe_RunningSpecs/Describe_Example.vb
Imports NSpec
Imports NUnit.Framework
Imports NSpec.Tests
Imports NSpec.Tests.WhenRunningSpecs
Imports FluentAssertions

Namespace WhenRunningSpecs

    <TestFixture()>
    <Category("RunningSpecs")>
    Public Class DescribeExample
        Inherits when_running_specs

        Public Class SpecClass
            Inherits nspec

            Sub It_Changes_Status_After_Run()

            End Sub

            Sub It_Passes()

            End Sub

            Sub It_Fails()
                Throw New Exception()
            End Sub
        End Class

        <Test()>
        Public Sub Execution_Status_Changes_After_Run()
            Run(GetType(SpecClass))

            Dim ex = TheExample("It Changes Status After Run")

            ex.HasRun.Should().BeTrue()
        End Sub

        <Test()>
        Public Sub Passing_Status_Is_Passed_When_It_Succeeds()
            Run(GetType(SpecClass))

            TheExample("It Passes").ShouldHavePassed()
        End Sub

        <Test()>
        Public Sub Passing_Status_Is_Not_Passed_When_It_Fails()
            Run(GetType(SpecClass))

            TheExample("It Fails").ShouldHaveFailed()
        End Sub

        Public Class SpecClassWithAnonymousLambdas
            Inherits nspec

            Sub Describe_Specs_In_VB_With_Anonymous_Lambdas()
                context("Some context with an anonymous Lambda") =
                    Sub()
                        it("has an anonymous lambda") = Sub()
                                                        End Sub
                    End Sub
            End Sub
        End Class

        <Test()>
        Public Sub Finds_And_Runs_Three_Class_Level_Examples()
            Run(GetType(SpecClass))

            TheExampleCount().Should().Be(3)
        End Sub

        <Test()>
        Public Sub Finds_Only_One_Example_Ignoring_Anonymous_Lambdas()
            Run(GetType(SpecClassWithAnonymousLambdas))

            TheExampleCount().Should().Be(1)
        End Sub

    End Class

End Namespace

